pkg/platforms/platforms.go

```diff
@@ -0,0 +1,26 @@
+package platforms
+
+import (
+	cplatforms "github.com/containerd/containerd/platforms"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type allPlatformsWithPreferenceMatcher struct {
+	preferred cplatforms.MatchComparer
+}
+
+// AllPlatformsWithPreference will return a platform matcher that matches all
+// platforms but will order platforms matching the preferred matcher first.
+func AllPlatformsWithPreference(preferred cplatforms.MatchComparer) cplatforms.MatchComparer {
+	return allPlatformsWithPreferenceMatcher{
+		preferred: preferred,
+	}
+}
+
+func (c allPlatformsWithPreferenceMatcher) Match(_ ocispec.Platform) bool {
+	return true
+}
+
+func (c allPlatformsWithPreferenceMatcher) Less(p1, p2 ocispec.Platform) bool {
+	return c.preferred.Less(p1, p2)
```

Comment:
If this is to live in a `pkg` package, we should make sure it's robust for any `preferred` matcher, including ones whose `Less` implementation may not necessarily sort platforms which `Match` less than platforms it would not.
```suggestion
	m1, m2 := c.preferred.Match(p1), c.preferred.Match(p2)
	if m1 && m2 {
		return c.preferred.Less(p1, p2)
	}
	return m1 // Not totally-ordered
```